tasklist/sqlite3_engine.py

```diff
@@ -0,0 +1,316 @@
+                        WHERE status & ?
+                    """),
+                    (st_aborted, datetime_to_epoch(now), st_doing)
+                )
+                # Reset scheduled task to default
+                c.execute(
+                    dedent("""
+                        UPDATE tasks SET status = ?
+                        WHERE status & ?
+                    """),
+                    (st_default, st_scheduled)
+                )
+        except sqlite3.Error as e:
+            logger.exception(str(e))
+            raise Exception("Could not recover tasks.")
+
+    def list_to_do(self, status, now, redo=False):
+        query = dedent("""
+                SELECT
+                    id, worker_name, start_datetime, stop_datetime,
+                    status, output, options, redo_interval, expire
+                FROM tasks
+                WHERE
+        """)
+        if redo:
+            query += dedent("""
+                    redo_interval > 0
+                    AND (start_datetime + redo_interval) < ?
+                    AND status & ?
+            """)
+        else:
+            query += dedent("""
+                    start_datetime < ?
+                    AND status & ?
+            """)
+
+        try:
+            with self.conn:
+                c = self.conn.cursor()
+                c.execute(query, (datetime_to_epoch(now), status,))
+                for r in c.fetchall():
+                    try:
+                        yield Task(
+                            id=r[0],
+                            worker_name=r[1],
+                            start_datetime=epoch_to_datetime(r[2]),
+                            stop_datetime=epoch_to_datetime(r[3]),
+                            status=r[4],
+                            output=r[5],
+                            options=json.loads(r[6]),
+                            redo_interval=r[7],
+                            expire=r[8]
+                        )
+                    except ValueError as e:
+                        logger.exception(str(e))
+                        logger.error("Could not convert row to Task.")
+                        logger.debug(r)
+                        continue
+
+        except sqlite3.Error as e:
+            logger.exception(str(e))
+            raise Exception("Could not get task list.")
+
+    def purge(self, status, now):
+        try:
+            with self.conn:
+                c = self.conn.cursor()
+                c.execute(
+                    dedent("""
+                    DELETE FROM tasks
+                    WHERE
+                        redo_interval = 0
+                        AND (stop_datetime + expire) < ?
+                        AND status & ?
+                    """),
+                    (datetime_to_epoch(now), status)
+                )
+        except sqlite3.Error as e:
+            logger.exception(str(e))
+            raise Exception("Could not purge task list.")
```

Comment:
Actually, I think we should be agnostic of storage engine in the upper layer: I don't want to deal with `sqlite3` errors in `taskmanager.py` because we could add support of another storage engine in the future, like Postgres. About less specific `Exception` raised, it could be replaced with a new `StorageEngineError` exception, what do you think ?

Reply:
I think that it does not matter much as long as the exception is not catched downstream. If it were to be catched, then I also agree that a specific and backend-agnostic exception type would be better.

Reply:
3c6dec1168179a1bff33d1381a380d906d2a4d2b should addresses it I hope.